caddy_cache.json

```diff
@@ -9,19 +9,44 @@
 					"listen": [":80"],
 					"routes": [
 						{
+							"match": [{
+								"path": ["/cache-default"]
+							}],
 							"handle": [
 								{
-									"handler": "cache"
+									"handler": "cache",
+									"default_ttl": 60
 								},
 								{
 									"handler": "static_response",
 									"body": "Hello there!"
 								}
 							]
+						},
+						{
+							"match": [{
+								"path": ["/cache-header"]
+							}],
+							"handle": [
+								{
+									"handler": "cache",
+									"olric": {
+										"bind_port": 6788,
```

Comment:
This is necessary to prevent port conflicts. Currently every instance of the cache module starts its own Olric instance. I don't know if it's a good practice or not (cc @buraksezer)

Reply:
That's kind of what this comment was getting at: https://github.com/caddyserver/cache-handler/pull/5/files#diff-11c97206e17bbda59c9250af3f2585b8L66 

Does it make sense to ever have more than one cache? If not, we wouldn't even need a UsagePool: just make a single package-scope variable like `var olricCache = ...`; however, we might need a UsagePool if the actual cache _storage_ can't be separated from its config. In that case, we just use a UsagePool so that the old cache is deallocated when the Caddy config is changed, and a new cache takes its place. Of course, if we need to repopulate the cache... that could be an issue.

I'd really like a way to have one cache _storage_ so that we don't have to keep emptying it for each reload. CertMagic, which manages the certificates, has a similar problem that I've been meaning to get around to, in that there's no way to say "keep this piece of data, but discard that one, but keep this one, ...".

So the scope and number of instances of a cache probably should be discussed.

Reply:
I pushed a new commit implementing the following strategy:

* there is only one instance of Olric per Caddy server and the cache is kept between reloads: I also think that it's what most users expect
* if there is a config change, the instance of Olric is destroyed and recreated with the new configuration (the cache is cleared)
* I added the host and the method to the cache key to prevent poisoning the cache with resources from other virtual hosts (this need to be double checked)

It's far from being complete and optimal, but it's a good step to something usable.

> CertMagic, which manages the certificates, has a similar problem that I've been meaning to get around to, in that there's no way to say "keep this piece of data, but discard that one, but keep this one, ...".

Regarding this, couldn't we add a special bag in `caddy.Context` to store this kind of global objects? I don't think that it's necessary here (at least now), but it may help for other use cases such as the one with CertMagic.

Reply:
> if there is a config change, the instance of Olric is destroyed and recreated with the new configuration (the cache is cleared)

Some Caddy users push config changes very frequently via the API, so clearing the cache may not be desirable for them. Is there room for this being configurable? Default to destroying on reload, but option to preserve the cache across reloads?

Reply:
> if there is a config change, the instance of Olric is destroyed and recreated with the new configuration (the cache is cleared)

I think this can be dangerous. If the user reload the configuration concurrently on many servers, this can lead unexpected results. It depends on the Olric configuration. These values should be configured carefully: `writeQuorum`, `readQuorum`, `memberCountQuorum`. 

We probably need to find a way to keep cache instances between configuration reloads.  

> I added the host and the method to the cache key to prevent poisoning the cache with resources from other virtual hosts (this need to be double checked)

I'm not sure about this. By this way, redundant Caddy instances will hit the origin server instead of reading the value from a neighboring Caddy instance. Is this the desired behavior?

Reply:
@francislavoie

> Some Caddy users push config changes very frequently via the API, so clearing the cache may not be desirable for them.

The cache is recreated only if Olric's config changed. Not at every (unrelated) Caddy config change: https://github.com/caddyserver/cache-handler/pull/5/files#diff-11c97206e17bbda59c9250af3f2585b8R90

@buraksezer

> If the user reload the configuration concurrently on many servers, this can lead unexpected results.

Good point! What do you recommend to do? Maybe can we start an Olric instance when the module boots and that's all? If the user wants to change the configuration it has to restart the server(s) by itself. We can even detect attempt to change Olric's config in the `Validate` function and return an error in this case. This will also make the code simpler. And it will always be possible to implement a more subtle behavior in the future, if needed at some point.

> I'm not sure about this. By this way, redundant Caddy instances will hit the origin server instead of reading the value from a neighboring Caddy instance.

No that's not how it works. I added the **public** host (`http.Request.Host`, which is based on the `Host` HTTP header) to the cache key, not the internal hostname of the server. This is mandatory if we have only one cache because Caddy can serve multiple public websites with different domain names. For instance let's say that you have two websites served by the same Caddy server (which is very common):

* https://site1.com
* https://site2.com

If we don't add the host to the cache key, we have a cache poisoning problem: a first user visits `https://site1.com/foo`, this is stored in cache, the next user visits `https://site2.com/foo` (it's a totally unrelated page), but we wrongly serve the page of `site1.com`.

Reply:
Yeah, this is a little tricky. Caddy already has tools for managing global state through reloads (see [UsagePool](https://pkg.go.dev/github.com/caddyserver/caddy/v2#UsagePool), for example) - I think the primary question is, how to transition between configs with Olric? I'd need to know more about how Olric works in order to help decide that.

@buraksezer and @dunglas, would you like invites to our developer Slack? I'm OK with continuing the discussion on GitHub but if it's helpful to have tighter feedback loops and quicker roundtrips, you'd be welcome to our Slack. Just let me know which email to invite (e.g. the ones in your profile, or whatever).

Reply:
Sure, Slack will be easier.

Reply:
> Maybe can we start an Olric instance when the module boots and that's all? If the user wants to change the configuration he has to restart the server(s) by himself. 

Olric currently doesn't support reloading configuration at runtime. We may implement that feature but I don't know whether hashicorp/memberlist supports this feature or not. 

In addition, some of the variables are impossible to change such as `PartitionCount` or `BindAddr/Port`. I also don't know what happens we change `ReplicaCount`, 'Write/ReadQuorum' variables. Olric is designed to run with static configuration. 

I think that it's possible to reload some part of the configuration. Such as:

* CacheConfig
* DMapCacheConfig
* Timeouts
* MaxInuse
* EvictionPolicy

I don't know what the cost is to implement hot-reload feature but I'll analyze it.

Olric instance should not be restarted through an unrelated configuration change. Additionally, restarting an Olric cluster is a non-trivial task. Look at this simple cases:

* If you have `ReplicaCount=2`, you can keep all the data if you restart the servers in some order. 
* If you have `ReplicaCount=1`, all the data on the restarted instance is gone. 
* When a new node join(because you restarted it) the cluster, some of the partitions and backups are migrated to the new node. If the `PartitionCount` variable is not selected correctly, these migrations may lead latency spikes. Maximum partition size should be around 100MB.

I didn't mention split brain scenarios yet. Olric implements something called majority quorum but if you don't care about consistency, just set `MemberCountQuorum=1` to disable it. 

We have to document this kind of things very clearly. I'm going to test it in every aspect and fix the bugs before a release.